src/ApiException.php

```diff
@@ -78,6 +78,50 @@ public function getStatus()
         return $this->status;
     }
 
+    // Returns null if metadata does not contain error info, or return containsErrorInfo() array
+    // if the metadata does contain error info.
+    public static function decodeMetadataErrorInfo($metadata)
+    {
+        // For the createFromRpcStatus() method, $metadata is an object.
+        if (is_object($metadata)) {
+            $metadataRpcStatus = Serializer::decodeAnyMessages($metadata);
+            $detailsRpcStatus = self::containsErrorInfo($metadataRpcStatus);
+            if ($detailsRpcStatus['containsErrorInfo']) {
+                return  $detailsRpcStatus;
+            }
+        } elseif (is_Array($metadata)) {
+            // For REST-based responses, the $metadata does not need to be decoded.
+            $detailsRest = self::containsErrorInfo($metadata);
+            if ($detailsRest['containsErrorInfo']) {
+                return $detailsRest;
+            }
+        }
+        $metadataGrpc = Serializer::decodeMetadata($metadata);
+        $detailsGrpc = self::containsErrorInfo($metadataGrpc);
+        return $detailsGrpc['containsErrorInfo'] ? $detailsGrpc : null;
+    }
+    
+    public function getDomain()
+    {
+        $metadata = $this->metadata;
+        return isset(self::decodeMetadataErrorInfo($metadata)['domain'])
+        ? self::decodeMetadataErrorInfo($metadata)['domain'] : null;
```

Comment:
Instead of invoking this twice, should the result just be captured in a variable then evaluated in the ternary operator? This comment goes for the other accessors.

Reply:
Makes sense - updated!